flink-core/src/main/java/org/apache/flink/api/java/ClosureCleaner.java

```diff
@@ -169,7 +175,17 @@ && needsRecursion(f, fieldObject)) {
     private static boolean needsRecursion(Field f, Object fo) {
         return (fo != null
                 && !Modifier.isStatic(f.getModifiers())
-                && !Modifier.isTransient(f.getModifiers()));
+                && !Modifier.isTransient(f.getModifiers())
+                && !canBeSerialized(fo));
+    }
+
+    private static boolean canBeSerialized(Object o) {
+        try {
+            InstantiationUtil.serializeObject(o);
+            return true;
+        } catch (Exception e) {
```

Comment:
FWIW, it might be good idea to log the exception here (debug level) or catch directly NotSerializableException, otherwise this might hide some other exceptions.